Sources/Algorithms/CommonPrefix.swift

```diff
@@ -0,0 +1,350 @@
+
+//===----------------------------------------------------------------------===//
+// Sequence.commonPrefix(with:)
+//===----------------------------------------------------------------------===//
+
+extension Sequence {
+  @inlinable
+  public func commonPrefix<Other: Sequence>(
+    with other: Other,
+    by areEquivalent: (Element, Other.Element) throws -> Bool
+  ) rethrows -> [Element] {
+    var iterator = makeIterator()
+    var otherIterator = other.makeIterator()
+    var result: [Element] = []
+    
+    while let next = iterator.next(),
+          let otherNext = otherIterator.next(),
+          try areEquivalent(next, otherNext)
+    {
+      result.append(next)
+    }
+    
+    return result
+  }
+}
+
+extension Sequence where Element: Equatable {
+  @inlinable
+  public func commonPrefix<Other: Sequence>(
+    with other: Other
+  ) -> CommonPrefix<Self, Other> where Other.Element == Element {
+    CommonPrefix(base: self, other: other, areEquivalent: ==)
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// LazySequenceProtocol.commonPrefix(with:)
+//===----------------------------------------------------------------------===//
+
+extension LazySequenceProtocol {
+  @inlinable
+  public func commonPrefix<Other: Sequence>(
+    with other: Other,
+    by areEquivalent: @escaping (Element, Other.Element) -> Bool
+  ) -> CommonPrefix<Self, Other> {
+    CommonPrefix(base: self, other: other, areEquivalent: areEquivalent)
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Collection.commonPrefix(with:)
+//===----------------------------------------------------------------------===//
+
+extension Collection {
+  @inlinable
+  public func commonPrefix<Other: Sequence>(
+    with other: Other,
+    by areEquivalent: (Element, Other.Element) throws -> Bool
+  ) rethrows -> SubSequence {
+    let endIndex = endIndex
+    
+    var index = startIndex
+    var iterator = other.makeIterator()
+    
+    while index != endIndex,
+          let next = iterator.next(),
+          try areEquivalent(self[index], next)
+    {
+      formIndex(after: &index)
+    }
+    
+    return self[..<index]
+  }
+}
+
+extension Collection where Element: Equatable {
+  @inlinable
+  public func commonPrefix<Other: Sequence>(
+    with other: Other
+  ) -> SubSequence where Other.Element == Element {
+    commonPrefix(with: other, by: ==)
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// LazyCollectionProtocol.commonPrefix(with:)
+//===----------------------------------------------------------------------===//
+
+extension LazyCollectionProtocol {
+  @inlinable
+  public func commonPrefix<Other: Sequence>(
+    with other: Other,
+    by areEquivalent: @escaping (Element, Other.Element) -> Bool
+  ) -> CommonPrefix<Self, Other> {
+    CommonPrefix(base: self, other: other, areEquivalent: areEquivalent)
+  }
```

Comment:
Same as in the `LazySequenceProtocol` section: shouldn't these be using `Elements` and `elements` instead of `Self` and `self`?